File: packages/sdk-go/reasons.go
package rollgate

// EvaluationReasonKind represents the category of reason for a flag evaluation.
type EvaluationReasonKind string

const (
	// ReasonOff indicates the flag is disabled.
	ReasonOff EvaluationReasonKind = "OFF"
	// ReasonTargetMatch indicates the user is in the target users list.
	ReasonTargetMatch EvaluationReasonKind = "TARGET_MATCH"
	// ReasonRuleMatch indicates the user matched a targeting rule.
	ReasonRuleMatch EvaluationReasonKind = "RULE_MATCH"
	// ReasonFallthrough indicates no rules matched, using default rollout.
	ReasonFallthrough EvaluationReasonKind = "FALLTHROUGH"
	// ReasonError indicates an error occurred during evaluation.
	ReasonError EvaluationReasonKind = "ERROR"
	// ReasonUnknown indicates the flag was not found or reason is unknown.
	ReasonUnknown EvaluationReasonKind = "UNKNOWN"
)

// EvaluationErrorKind represents types of errors during evaluation.
type EvaluationErrorKind string

const (
	// ErrorFlagNotFound indicates the flag key does not exist.
	ErrorFlagNotFound EvaluationErrorKind = "FLAG_NOT_FOUND"
	// ErrorMalformedFlag indicates the flag configuration is invalid.
	ErrorMalformedFlag EvaluationErrorKind = "MALFORMED_FLAG"
	// ErrorUserNotSpecified indicates no user context was provided.
	ErrorUserNotSpecified EvaluationErrorKind = "USER_NOT_SPECIFIED"
	// ErrorClientNotReady indicates the SDK client is not initialized.
	ErrorClientNotReady EvaluationErrorKind = "CLIENT_NOT_READY"
	// ErrorException indicates an unexpected error occurred.
	ErrorException EvaluationErrorKind = "EXCEPTION"
)

// EvaluationReason explains why a flag evaluated to a particular value.
type EvaluationReason struct {
	// Kind is the general category of the reason.
	Kind EvaluationReasonKind `json:"kind"`
	// RuleID is the unique identifier of the matched rule (for RULE_MATCH).
	RuleID string `json:"ruleId,omitempty"`
	// RuleIndex is the 0-based index of the matched rule (for RULE_MATCH).
	RuleIndex int `json:"ruleIndex,omitempty"`
	// InRollout indicates whether the user was included in the rollout percentage.
	InRollout bool `json:"inRollout,omitempty"`
	// ErrorKind is the specific error type if Kind is ERROR.
	ErrorKind EvaluationErrorKind `json:"errorKind,omitempty"`
}

// EvaluationDetail contains the full result of a flag evaluation.
type EvaluationDetail[T any] struct {
	// Value is the evaluated flag value.
	Value T `json:"value"`
	// Reason explains why this value was returned.
	Reason EvaluationReason `json:"reason"`
	// VariationIndex is the index of the selected variation (for multi-variate flags).
	VariationIndex int `json:"variationIndex,omitempty"`
	// VariationID is the ID of the selected variation (for multi-variate flags).
	VariationID string `json:"variationId,omitempty"`
}

// BoolEvaluationDetail is an alias for EvaluationDetail[bool] for convenience.
type BoolEvaluationDetail = EvaluationDetail[bool]

// Helper functions to create common reasons

// OffReason creates a reason for a disabled flag.
func OffReason() EvaluationReason {
	return EvaluationReason{Kind: ReasonOff}
}

// TargetMatchReason creates a reason for a target user match.
func TargetMatchReason() EvaluationReason {
	return EvaluationReason{Kind: ReasonTargetMatch}
}

// RuleMatchReason creates a reason for a rule match.
func RuleMatchReason(ruleID string, ruleIndex int, inRollout bool) EvaluationReason {
	return EvaluationReason{
		Kind:      ReasonRuleMatch,
		RuleID:    ruleID,
		RuleIndex: ruleIndex,
		InRollout: inRollout,
	}
}

// FallthroughReason creates a reason for fallthrough to default rollout.
func FallthroughReason(inRollout bool) EvaluationReason {
	return EvaluationReason{
		Kind:      ReasonFallthrough,
		InRollout: inRollout,
	}
}

// ErrorReason creates a reason for an error.
func ErrorReason(errorKind EvaluationErrorKind) EvaluationReason {
	return EvaluationReason{
		Kind:      ReasonError,
		ErrorKind: errorKind,
	}
}

// UnknownReason creates a reason for an unknown flag.
func UnknownReason() EvaluationReason {
	return EvaluationReason{Kind: ReasonUnknown}
}

File: packages/sdk-core/src/reasons.ts
/**
 * Evaluation Reasons for Rollgate SDKs
 *
 * Provides detailed information about why a flag evaluated to a particular value.
 * Compatible with LaunchDarkly's evaluation reason format.
 */

/**
 * The kind of reason that caused a particular flag evaluation result.
 */
export type EvaluationReasonKind =
  | "OFF" // Flag is disabled
  | "TARGET_MATCH" // User is in the target users list
  | "RULE_MATCH" // User matched a targeting rule
  | "FALLTHROUGH" // No rules matched, using default rollout
  | "ERROR" // An error occurred during evaluation
  | "UNKNOWN"; // Flag not found or unknown reason

/**
 * Detailed evaluation reason explaining why a flag returned a particular value.
 */
export interface EvaluationReason {
  /** The general category of the reason */
  kind: EvaluationReasonKind;

  /** The unique identifier of the matched rule (for RULE_MATCH) */
  ruleId?: string;

  /** The 0-based index of the matched rule (for RULE_MATCH) */
  ruleIndex?: number;

  /** Whether this evaluation is part of an experiment (for RULE_MATCH or FALLTHROUGH) */
  inExperiment?: boolean;

  /** The specific error type if kind is ERROR */
  errorKind?: EvaluationErrorKind;

  /** Whether the user was included in the rollout percentage */
  inRollout?: boolean;
}

/**
 * Types of errors that can occur during evaluation.
 */
export type EvaluationErrorKind =
  | "FLAG_NOT_FOUND" // The flag key does not exist
  | "MALFORMED_FLAG" // The flag configuration is invalid
  | "USER_NOT_SPECIFIED" // No user context was provided for targeting
  | "CLIENT_NOT_READY" // The SDK client is not initialized
  | "EXCEPTION"; // An unexpected error occurred

/**
 * Detailed evaluation result including the value and reason.
 */
export interface EvaluationDetail<T = boolean> {
  /** The evaluated flag value */
  value: T;

  /** The reason for this evaluation result */
  reason: EvaluationReason;

  /** Index of the selected variation (for multi-variate flags) */
  variationIndex?: number;

  /** ID of the selected variation (for multi-variate flags) */
  variationId?: string;
}

/**
 * Create a reason for a disabled flag.
 */
export function offReason(): EvaluationReason {
  return { kind: "OFF" };
}

/**
 * Create a reason for a target user match.
 */
export function targetMatchReason(): EvaluationReason {
  return { kind: "TARGET_MATCH" };
}

/**
 * Create a reason for a rule match.
 */
export function ruleMatchReason(
  ruleId: string,
  ruleIndex: number,
  inRollout: boolean = true,
): EvaluationReason {
  return {
    kind: "RULE_MATCH",
    ruleId,
    ruleIndex,
    inRollout,
  };
}

/**
 * Create a reason for fallthrough to default rollout.
 */
export function fallthroughReason(inRollout: boolean = true): EvaluationReason {
  return {
    kind: "FALLTHROUGH",
    inRollout,
  };
}

/**
 * Create a reason for an error.
 */
export function errorReason(errorKind: EvaluationErrorKind): EvaluationReason {
  return {
    kind: "ERROR",
    errorKind,
  };
}

/**
 * Create a reason for an unknown flag.
 */
export function unknownReason(): EvaluationReason {
  return { kind: "UNKNOWN" };
}
